Fix/citizen data command parsing

mkienenb opened the pull request:
# Changes proposed in this pull request:
- Change tab completion for citizen argument type to return citizen ids not colony ids.
- Change tab completion for citizen argument type to limit citizens returned by a previously-specific colony argument value.
- Fix problem with commands remembering state of parameters from the past time they were used.
- Added missing "rs" navigation menu under main minecolonies menu.
- Renamed PLAYER argument type to ONLINE_PLAYER and added all PLAYER argument type.
- Changed SCAN and RTP commands to use ONLINE_PLAYER.

Review please

mkienenb committed:
Fix  tab completion showing ids for citizen.

mkienenb committed:
Use previous colony value to determine what citizens to return.

mkienenb committed:
Fix codacy warning.

mkienenb committed:
Add missing RS menu item to main navigation menu.

mkienenb committed:
Fix code warnings.

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +201,32 @@ public int allowedSpaceCount()
+        {
+            final ActionMenu actionMenu = (ActionMenu)menu;
+            // Try to find a valid colony value.
+            // TODO: doesn't check subarguments but should only check arguments for parents of this argument.
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +201,32 @@ public int allowedSpaceCount()
+            final ActionMenu actionMenu = (ActionMenu)menu;
+            // Try to find a valid colony value.
+            // TODO: doesn't check subarguments but should only check arguments for parents of this argument.
+            // TODO: Also no guarantee that we've grabbed the right colony argument if the command has more than one.
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -84,9 +84,17 @@ public int allowedSpaceCount()
     }
 
-    private List<String> getCitizenNames()
+    private List<String> getCitizenNames(@Nullable final Colony colonyToUse)
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Make "getCitizenNames" a "static" method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2325)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +201,32 @@ public int allowedSpaceCount()
     }
 
-    private List<String> getCitizenTabCompletions(final String potentialArgumentValue)
+    private List<String> getCitizenTabCompletions(@NotNull final TreeNode<IMenu> actionMenuTreeNode, final String potentialArgumentValue)
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "getCitizenTabCompletions" is 18 which is greater than 15 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 44 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +201,32 @@ public int allowedSpaceCount()
-    private List<String> getCitizenTabCompletions(final String potentialArgumentValue)
+    private List<String> getCitizenTabCompletions(@NotNull final TreeNode<IMenu> actionMenuTreeNode, final String potentialArgumentValue)
     {
         // TODO: see if we can figure out what citizen we are looking at as the default tab completion.
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

BuildTeamLinter commented:
SonarQube analysis reported 65 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 54 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 7 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [CommandEntryPointNew.java#L292](https://github.com/Minecolonies/minecolonies/blob/3bbd3d5946ce9e3493307e01f1a85e9fdbb07075/src/main/java/com/minecolonies/coremod/commands/CommandEntryPointNew.java#L292): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/3bbd3d5946ce9e3493307e01f1a85e9fdbb07075/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/3bbd3d5946ce9e3493307e01f1a85e9fdbb07075/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/3bbd3d5946ce9e3493307e01f1a85e9fdbb07075/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/3bbd3d5946ce9e3493307e01f1a85e9fdbb07075/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/3bbd3d5946ce9e3493307e01f1a85e9fdbb07075/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/3bbd3d5946ce9e3493307e01f1a85e9fdbb07075/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/3bbd3d5946ce9e3493307e01f1a85e9fdbb07075/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [data#](https://github.com/Minecolonies/minecolonies/blob/3bbd3d5946ce9e3493307e01f1a85e9fdbb07075/src/main/java/com/minecolonies/coremod/colony/requestsystem/data#): Add a 'package-info.java' file to document the 'data' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [init#](https://github.com/Minecolonies/minecolonies/blob/3bbd3d5946ce9e3493307e01f1a85e9fdbb07075/src/main/java/com/minecolonies/coremod/colony/requestsystem/init#): Add a 'package-info.java' file to document the 'init' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +201,32 @@ public int allowedSpaceCount()
+        {
+            final ActionMenu actionMenu = (ActionMenu) menu;
+            // Try to find a valid colony value.
+            // TODO: doesn't check subarguments but should only check arguments for parents of this argument.
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +201,32 @@ public int allowedSpaceCount()
+            final ActionMenu actionMenu = (ActionMenu) menu;
+            // Try to find a valid colony value.
+            // TODO: doesn't check subarguments but should only check arguments for parents of this argument.
+            // TODO: Also no guarantee that we've grabbed the right colony argument if the command has more than one.
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

BuildTeamLinter commented:
SonarQube analysis reported 64 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 54 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 7 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [CommandEntryPointNew.java#L292](https://github.com/Minecolonies/minecolonies/blob/174f8f188f4c1af13b0df1380350954a065f7508/src/main/java/com/minecolonies/coremod/commands/CommandEntryPointNew.java#L292): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/174f8f188f4c1af13b0df1380350954a065f7508/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/174f8f188f4c1af13b0df1380350954a065f7508/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/174f8f188f4c1af13b0df1380350954a065f7508/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/174f8f188f4c1af13b0df1380350954a065f7508/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/174f8f188f4c1af13b0df1380350954a065f7508/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/174f8f188f4c1af13b0df1380350954a065f7508/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/174f8f188f4c1af13b0df1380350954a065f7508/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [data#](https://github.com/Minecolonies/minecolonies/blob/174f8f188f4c1af13b0df1380350954a065f7508/src/main/java/com/minecolonies/coremod/colony/requestsystem/data#): Add a 'package-info.java' file to document the 'data' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [init#](https://github.com/Minecolonies/minecolonies/blob/174f8f188f4c1af13b0df1380350954a065f7508/src/main/java/com/minecolonies/coremod/colony/requestsystem/init#): Add a 'package-info.java' file to document the 'init' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

mkienenb committed:
Change PLAYER_TYPE to use all players.

Add ONLINE_PLAYER type which only works with online players.
Add missing @NotNull annotations.

mkienenb committed:
Change SCAN and RTP commands to use ONLINE_PLAYER

mkienenb committed:
Fix more whining about every method needing to be static.

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionMenuType.java:
```diff
@@ -37,7 +37,7 @@
             ForgePermissionNodes.RANDOM_TELEPORT,
             RandomTeleportCommand.class,
-            new ActionArgument("player", ActionArgumentType.PLAYER, ActionArgumentType.Is.OPTIONAL)
+            new ActionArgument("player", ActionArgumentType.ONLINE_PLAYER, ActionArgumentType.Is.OPTIONAL)
```
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "player" 6 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -67,12 +69,31 @@ public int allowedSpaceCount()
+    }
+
+    @NotNull
     private List<String> getColonyIdStrings()
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Make "getColonyIdStrings" a "static" method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2325)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -151,6 +209,7 @@ public int allowedSpaceCount()
+    @NotNull
     private List<String> getColonyTabCompletions(final String potentialArgumentValue)
     {
         // TODO: use the colony we are in as the default tab completion.
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +229,33 @@ public int allowedSpaceCount()
 
-    private List<String> getCitizenTabCompletions(final String potentialArgumentValue)
+    @NotNull
+    private List<String> getCitizenTabCompletions(@NotNull final TreeNode<IMenu> actionMenuTreeNode, final String potentialArgumentValue)
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "getCitizenTabCompletions" is 18 which is greater than 15 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 44 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +229,33 @@ public int allowedSpaceCount()
+    @NotNull
+    private List<String> getCitizenTabCompletions(@NotNull final TreeNode<IMenu> actionMenuTreeNode, final String potentialArgumentValue)
     {
         // TODO: see if we can figure out what citizen we are looking at as the default tab completion.
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -99,9 +128,34 @@ public int allowedSpaceCount()
+    }
+
+    @NotNull
     public List<String> getTabCompletions(@NotNull final MinecraftServer server,
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 7, down to the maximum allowed 6. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)

BuildTeamLinter commented:
SonarQube analysis reported 73 issues
* ![CRITICAL][CRITICAL] 5 critical
* ![MAJOR][MAJOR] 56 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 10 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [ActionArgumentType.java#L426](https://github.com/Minecolonies/minecolonies/blob/c052bf7f63c2c4998e8710154e2f323d384527fd/src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java#L426): Define a constant instead of duplicating this literal "[abandoned]" 4 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ActionMenuType.java#L106](https://github.com/Minecolonies/minecolonies/blob/c052bf7f63c2c4998e8710154e2f323d384527fd/src/main/java/com/minecolonies/coremod/commands/ActionMenuType.java#L106): Define a constant instead of duplicating this literal "colony" 16 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ActionMenuType.java#L115](https://github.com/Minecolonies/minecolonies/blob/c052bf7f63c2c4998e8710154e2f323d384527fd/src/main/java/com/minecolonies/coremod/commands/ActionMenuType.java#L115): Define a constant instead of duplicating this literal "citizen" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [CommandEntryPointNew.java#L292](https://github.com/Minecolonies/minecolonies/blob/c052bf7f63c2c4998e8710154e2f323d384527fd/src/main/java/com/minecolonies/coremod/commands/CommandEntryPointNew.java#L292): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/c052bf7f63c2c4998e8710154e2f323d384527fd/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/c052bf7f63c2c4998e8710154e2f323d384527fd/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/c052bf7f63c2c4998e8710154e2f323d384527fd/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/c052bf7f63c2c4998e8710154e2f323d384527fd/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/c052bf7f63c2c4998e8710154e2f323d384527fd/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/c052bf7f63c2c4998e8710154e2f323d384527fd/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

mkienenb committed:
Move state out of ActionMenu so commands don't remember previous execution states.

mkienenb committed:
Code cleanup

mkienenb committed:
Merge branch 'version/1.12' into fix/citizen-data-command-parsing

marchermans reviewed (dismissed):
+1 for the rs changes.

Raycoms commented:
@mkienenb can you fix the javadoc issues?

mkienenb committed:
Use constant for abandoned player name.

mkienenb committed:
Merge remote-tracking branch 'origin/version/1.12' into fix/citizen-data-command-parsing

mkienenb committed:
Merge branch 'fix/citizen-data-command-parsing' of github.com:Minecolonies/minecolonies into fix/citizen-data-command-parsing

mkienenb committed:
Fix broken javadoc

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/colonycommands/ColonyTeleportCommand.java:
```diff
@@ -74,12 +74,12 @@ public boolean canRankUseCommand(@NotNull final Colony colony, @NotNull final En
 
     @Override
-    public void execute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final ActionMenu actionMenu) throws CommandException
+    public void execute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final ActionMenuState actionMenuState) throws CommandException
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 25 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/colonycommands/MakeNotAutoDeletableCommand.java:
```diff
@@ -12,7 +12,7 @@
 import com.minecolonies.coremod.colony.ColonyManager;
 import com.minecolonies.coremod.commands.AbstractSingleCommand;
 import com.minecolonies.coremod.commands.ActionArgument;
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this unused import 'com.minecolonies.coremod.commands.ActionArgument'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +231,29 @@ public int allowedSpaceCount()
-        final List<String> citizenNumberStrings = getColonyIdStrings();
+        @Nullable Colony colony = null;
+        // Try to find a valid colony value.
+        // TODO: doesn't check subarguments but should only check arguments for parents of this argument.
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +231,29 @@ public int allowedSpaceCount()
+        @Nullable Colony colony = null;
+        // Try to find a valid colony value.
+        // TODO: doesn't check subarguments but should only check arguments for parents of this argument.
+        // TODO: Also no guarantee that we've grabbed the right colony argument if the command has more than one.
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +231,29 @@ public int allowedSpaceCount()
 
-    private List<String> getCitizenTabCompletions(final String potentialArgumentValue)
+    @NotNull
+    private List<String> getCitizenTabCompletions(@NotNull final ActionMenuState actionMenuState, final String potentialArgumentValue)
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "getCitizenTabCompletions" is 17 which is greater than 15 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 40 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java:
```diff
@@ -170,11 +231,29 @@ public int allowedSpaceCount()
+    @NotNull
+    private List<String> getCitizenTabCompletions(@NotNull final ActionMenuState actionMenuState, final String potentialArgumentValue)
     {
         // TODO: see if we can figure out what citizen we are looking at as the default tab completion.
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/CommandEntryPointNew.java:
```diff
@@ -287,7 +299,7 @@ private static void throwCommandUsageExceptionIfRequiredArgumentsAreNotProvided(
-                        if (!executionActionArgument.isValueSet())
+                        if (!actionMenuState.isValueSet(executionActionArgument))
                         {
                             if (null == badArgument)
```
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/CommandEntryPointNew.java:
```diff
@@ -267,15 +278,16 @@ public void execute(@NotNull final MinecraftServer server, @NotNull final IComma
+        actionCommand.execute(server, sender, actionMenuState);
     }
 
     private static void throwCommandUsageExceptionIfRequiredArgumentsAreNotProvided(@NotNull final TreeNode<IMenu> executionTreeNode, @NotNull final ActionMenu actionMenu,
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 34 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/CommandEntryPointNew.java:
```diff
@@ -442,13 +465,18 @@ private ParsingResult getTabCompletionsAndParsingHolders(
     }
 
     @NotNull
     private ParsingResult getTabCompletionsAndParsingHoldersForActionMenuTreeNode(
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/CommandEntryPointNew.java:
```diff
@@ -420,18 +432,29 @@ private ParsingResult getTabCompletionsAndParsingHolders(
+                    }
+                    else
+                    {
+                        actionMenuState = (ActionMenuState) null;
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this unnecessary cast to "ActionMenuState". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1905)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/CommandEntryPointNew.java:
```diff
@@ -442,13 +465,18 @@ private ParsingResult getTabCompletionsAndParsingHolders(
     @NotNull
     private ParsingResult getTabCompletionsAndParsingHoldersForActionMenuTreeNode(
                                            @NotNull final TreeNode<IMenu> actionMenuTreeNode,
+           // TODO: can parsedHolders, parsedActionArgumentList, and possibleActionCommands be merged into actionMenuState?
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/colonycommands/ShowColonyInfoCommand.java:
```diff
@@ -70,15 +70,15 @@ public String getCommandUsage(@NotNull final ICommandSender sender)
     }
 
-    public void execute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final ActionMenu actionMenu) throws CommandException
+    public void execute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final ActionMenuState actionMenuState) throws CommandException
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/generalcommands/RaidAllTonightCommand.java:
```diff
@@ -4,7 +4,7 @@
 import com.minecolonies.coremod.colony.ColonyManager;
 import com.minecolonies.coremod.commands.AbstractSingleCommand;
 import com.minecolonies.coremod.commands.ActionArgument;
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this unused import 'com.minecolonies.coremod.commands.ActionArgument'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionMenuState.java:
```diff
@@ -0,0 +1,210 @@
+        {
+            if (argumentName.equals(actionArgument.getName()))
+            {
+                if (ActionArgumentType.CITIZEN == actionArgument.getType())
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1066)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionMenuState.java:
```diff
@@ -0,0 +1,210 @@
+        {
+            if (argumentName.equals(actionArgument.getName()))
+            {
+                if (ActionArgumentType.PLAYER == actionArgument.getType())
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1066)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionMenuState.java:
```diff
@@ -0,0 +1,210 @@
+        {
+            if (argumentName.equals(actionArgument.getName()))
+            {
+                if (ActionArgumentType.COLONY == actionArgument.getType())
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1066)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionMenuState.java:
```diff
@@ -0,0 +1,210 @@
+                }
+            }
+        }
+        return null;
```
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Null is returned but a "Boolean" is expected. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2447)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionMenuState.java:
```diff
@@ -0,0 +1,210 @@
+    @NotNull private final ActionMenu actionMenu;
+    @NotNull private final Map<String, ArgumentState> argumentStateByActionArgumentNameMap = new HashMap<>();
+
+    private static class ArgumentState
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Add a constructor to the class, or provide default values. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1258)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionMenuState.java:
```diff
@@ -0,0 +1,210 @@
+        {
+            if (argumentName.equals(actionArgument.getName()))
+            {
+                if (ActionArgumentType.BOOLEAN == actionArgument.getType())
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1066)

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/colonycommands/ClaimChunksCommand.java:
```diff
@@ -2,7 +2,7 @@
-import com.minecolonies.coremod.commands.ActionMenu;
+import com.minecolonies.coremod.commands.ActionMenuState;
 import com.minecolonies.coremod.commands.IActionCommand;
 import com.minecolonies.coremod.util.TeleportToColony;
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this unused import 'com.minecolonies.coremod.util.TeleportToColony'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

BuildTeamLinter commented:
SonarQube analysis reported 120 issues
* ![CRITICAL][CRITICAL] 7 critical
* ![MAJOR][MAJOR] 90 major
* ![MINOR][MINOR] 11 minor
* ![INFO][INFO] 12 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [ActionMenuType.java#L106](https://github.com/Minecolonies/minecolonies/blob/5f9f7bb9ece5d54dcb44896fdba792cb13b9584a/src/main/java/com/minecolonies/coremod/commands/ActionMenuType.java#L106): Define a constant instead of duplicating this literal "colony" 16 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ActionMenuType.java#L115](https://github.com/Minecolonies/minecolonies/blob/5f9f7bb9ece5d54dcb44896fdba792cb13b9584a/src/main/java/com/minecolonies/coremod/commands/ActionMenuType.java#L115): Define a constant instead of duplicating this literal "citizen" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ColonyTeleportCommand.java#L94](https://github.com/Minecolonies/minecolonies/blob/5f9f7bb9ece5d54dcb44896fdba792cb13b9584a/src/main/java/com/minecolonies/coremod/commands/colonycommands/ColonyTeleportCommand.java#L94): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![CRITICAL][CRITICAL] [ColonyTeleportCommand.java#L112](https://github.com/Minecolonies/minecolonies/blob/5f9f7bb9ece5d54dcb44896fdba792cb13b9584a/src/main/java/com/minecolonies/coremod/commands/colonycommands/ColonyTeleportCommand.java#L112): Define a constant instead of duplicating this literal "You are not allowed to do this" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/5f9f7bb9ece5d54dcb44896fdba792cb13b9584a/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/5f9f7bb9ece5d54dcb44896fdba792cb13b9584a/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/5f9f7bb9ece5d54dcb44896fdba792cb13b9584a/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/5f9f7bb9ece5d54dcb44896fdba792cb13b9584a/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/5f9f7bb9ece5d54dcb44896fdba792cb13b9584a/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/5f9f7bb9ece5d54dcb44896fdba792cb13b9584a/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

mkienenb committed:
Improve javadoc.

mkienenb committed:
Remove unused imports.

mkienenb committed:
Initialize ArgumentState with null value for parameters.

Supress null returned for Boolean warning.

mkienenb committed:
Clean up BuildLinter warning.

BuildTeamLinter commented on src/main/java/com/minecolonies/coremod/commands/ActionMenuState.java:
```diff
@@ -0,0 +1,218 @@
+     * But in this case the rule does not apply because
+     * We are returning null to indicate that no boolean argument value was set.
+     */
+    @SuppressWarnings({"squid:S2447"})
```
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Suppressing the 'squid:S2447' warning is not allowed [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)

BuildTeamLinter commented:
SonarQube analysis reported 113 issues
* ![CRITICAL][CRITICAL] 6 critical
* ![MAJOR][MAJOR] 89 major
* ![MINOR][MINOR] 5 minor
* ![INFO][INFO] 13 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [ActionMenuType.java#L106](https://github.com/Minecolonies/minecolonies/blob/abfeaaf9f4d0960514d1b5dc9efc6e8e3f93cd49/src/main/java/com/minecolonies/coremod/commands/ActionMenuType.java#L106): Define a constant instead of duplicating this literal "colony" 16 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ActionMenuType.java#L115](https://github.com/Minecolonies/minecolonies/blob/abfeaaf9f4d0960514d1b5dc9efc6e8e3f93cd49/src/main/java/com/minecolonies/coremod/commands/ActionMenuType.java#L115): Define a constant instead of duplicating this literal "citizen" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ColonyTeleportCommand.java#L94](https://github.com/Minecolonies/minecolonies/blob/abfeaaf9f4d0960514d1b5dc9efc6e8e3f93cd49/src/main/java/com/minecolonies/coremod/commands/colonycommands/ColonyTeleportCommand.java#L94): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![CRITICAL][CRITICAL] [ColonyTeleportCommand.java#L112](https://github.com/Minecolonies/minecolonies/blob/abfeaaf9f4d0960514d1b5dc9efc6e8e3f93cd49/src/main/java/com/minecolonies/coremod/commands/colonycommands/ColonyTeleportCommand.java#L112): Define a constant instead of duplicating this literal "You are not allowed to do this" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/abfeaaf9f4d0960514d1b5dc9efc6e8e3f93cd49/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/abfeaaf9f4d0960514d1b5dc9efc6e8e3f93cd49/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/abfeaaf9f4d0960514d1b5dc9efc6e8e3f93cd49/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/abfeaaf9f4d0960514d1b5dc9efc6e8e3f93cd49/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/abfeaaf9f4d0960514d1b5dc9efc6e8e3f93cd49/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/abfeaaf9f4d0960514d1b5dc9efc6e8e3f93cd49/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'
